lib/transactional_update.pm

```diff
@@ -0,0 +1,181 @@
+# SUSE's openQA tests
+#
+# Copyright © 2017 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+package transactional_updates;
+
+use base Exporter;
+use Exporter;
+
+use strict;
+use warnings;
+
+our @EXPORT = qw(
+  check_reboot_changes
+  is_transactional_updates_system
+  microos_login
+  process_reboot
+  rpmver
+  trup_call
+  trup_install
+);
+
+# Reboot if there's a diff between the current FS and the new snapshot
+sub check_reboot_changes {
+    my $change_expected = shift // 1;
+
+    # Compare currently mounted and default subvolume
+    my $time    = time;
+    my $mounted = "mnt-$time";
+    my $default = "def-$time";
+    assert_script_run "mount | grep 'on / ' | egrep -o 'subvolid=[0-9]*' | cut -d'=' -f2 > $mounted";
+    assert_script_run "btrfs su get-default / | cut -d' ' -f2 > $default";
+    my $change_happened = script_run "diff $mounted $default";
+
+    # If changes are expected check that default subvolume changed
+    die "Error during diff" if $change_happened > 1;
+    die "Change expected: $change_expected, happeed: $change_happened" if $change_expected != $change_happened;
+
+    # Reboot into new snapshot
+    process_reboot 1 if $change_happened;
+}
+
+# Check if the SUT uses transactional-updates (read-only filesystem, rebootmgr)
+sub is_transactional_updates_system {
+    return (get_var('SYSTEM_ROLE') =~ /kubeadm|microos|serverro/);
+}
+
+# Assert login prompt and login as root
+sub microos_login {
```

Comment:
My idea is that this lib is a general one for the specific kubic and caasp.
I extracted this function since it is a common dependency, to avoid dependency lock.
Any better idea?

Reply:
Please explain what you mean by dependency lock.

Just saying that transactional lib should have common functions used only for transactional tests.

microos_login and process_reboot are wrappers handling caasp/kubic differences. 
For example `assert_screen 'linux-login-casp'` - is checking ssh keys in login prompt. You can't reuse that on TW/SLES. (if you create common needle it will disable ssh check)

If you use that on TW/SLES you will have to add sles wrappers to caasp wrappers - it might become spagetti. We probably have common functions handling reboot on sles/tw already.

Maybe create new internal sub in transactional (no need to split reboot/login here):
```perl
# lib/transactional.pm
sub _process_reboot {
    my $trigger = shift // 0;
    if (is_caasp) {
        caasp_reboot $trigger;
        caasp_login;
    } else {
        # this might be already in some sub...somewhere  ..power_action_utils.pm?
        power_action('reboot', observe => !$trigger, keepconsole => 1);
        assert_screen 'LOGIN-NEEDLE', 150;
        select_console 'root-console';
        assert_script_run 'clear';
    }
}
```

Reply:
I think you don't need to import `transactional` in `caasp` lib.

I never tried "cross dependency" - don't know if perl has problem with it.